### PR TITLE
feat(android): check for microphone permission so ongoing service can start

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 1]
```diff
@@ -95,26 +100,48 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
-            };
+
+                return true;
+            }
+        };
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            permissionsList.add(POST_NOTIFICATIONS);
 
             JitsiMeetActivityDelegate.requestPermissions(
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:120]: Don't immediately check the permissions here, do it only if the array is not empty.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 2]
```diff
@@ -95,26 +100,48 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
+
+        String[] permissionsArray = new String[ permissionsList.size() ];
+        permissionsList.toArray( permissionsArray );
+
+        for (String permission : permissionsList) {
+            boolean checkPermission
+                = ContextCompat.checkSelfPermission(context, permission)
+                == PackageManager.PERMISSION_GRANTED;
+
+            if (checkPermission) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:140]: This is wrong, it will pass if one of the permissions is granted but not the other one. Don;t check, just request them direcrtly.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 3]
```diff
@@ -132,7 +159,7 @@ public void onCreate() {
             stopSelf();
             JitsiMeetLogger.w(TAG + " Couldn't start service, notification is null");
         } else {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:162]: Why is this change necessary?

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 4]
```diff
@@ -32,13 +33,17 @@
...
 import android.os.IBinder;
 
+import androidx.core.content.ContextCompat;
 import androidx.localbroadcastmanager.content.LocalBroadcastManager;
 
 import com.facebook.react.modules.core.PermissionListener;
 
 import org.jitsi.meet.sdk.log.JitsiMeetLogger;
 
+import java.lang.reflect.Array;
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:43]: IS this import needed?

Comment (Calinteodor) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:43]: Leftover.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java (hunk 3 above)
Comment (Calinteodor) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:162]: Because minimum is now 30 for FOREGROUND_SERVICE_TYPE_MICROPHONE.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 5]
```diff
@@ -95,26 +99,43 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            permissionsList.add(RECORD_AUDIO);
+        }
+
+        String[] permissionsArray = new String[ permissionsList.size() ];
+        permissionsList.toArray( permissionsArray );
+
+        if (permissionsList.size() > 0) {
+
+            for (String permission : permissionsList) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:131]: Don't do a loop, you can pass the array and they are requested simultaneously.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 6]
```diff
@@ -95,26 +99,43 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
-                    }
-
-                    return true;
+        List<String> permissionsList = new ArrayList<String>();
+
+
+        PermissionListener listener = new PermissionListener() {
+            @Override
+            public boolean onRequestPermissionsResult(int i, String[] strings, int[] results) {
+                if (results.length > 0 && results[0] == PackageManager.PERMISSION_GRANTED) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:108]: Check all the results, not just the first.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 7]
```diff
@@ -95,26 +99,39 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
-                        doLaunch(context, extraData);
-                    }
-
-                    return true;
+        List<String> permissionsList = new ArrayList<>();
+
+        PermissionListener listener = new PermissionListener() {
+            @Override
+            public boolean onRequestPermissionsResult(int i, String[] strings, int[] results) {
+                if (results.length > 0 && results[i] == PackageManager.PERMISSION_GRANTED) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:107]: This needs to loop through all the results.

It seems like we're only checking the i-th result.

Does this get called multiple times?

Since we get a results array we need to check all of them once they are all resolved and only then launch.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 8]
```diff
@@ -95,24 +101,48 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
+                POST_NOTIFICATIONS_PERMISSIONS_REQUEST_CODE,
                 listener
             );
+
+            String[] permissionsArray = new String[ permissionsList.size() ];
+            permissionsArray = permissionsList.toArray( permissionsArray );
+
+
+            if (permissionsArray.length > 0) {
+                boolean permissionsResult
```
Comment (Calinteodor) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:135]: @saghul I reworked the logic a bit.
I suggest we use the permission request listener only for POST_NOTIFICATIONS. That's it.
Both RECORD_AUDIO and POST_NOTIFICATIONS need to be granted.
We launch the service under the API >= 33 if all of the above check and we just launch for API < 33.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 9]
```diff
@@ -132,10 +154,16 @@ public void onCreate() {
             stopSelf();
             JitsiMeetLogger.w(TAG + " Couldn't start service, notification is null");
         } else {
+            if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.Q) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:157]: This and the next check both cover Q. Which one should Q execute? Also make it if-else if-else please.

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 10]
```diff
@@ -95,26 +98,45 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
-                    if (results.length > 0 && results[0] == PackageManager.PERMISSION_GRANTED) {
+        List<String> permissionsList = new ArrayList<>();
+
+        PermissionListener listener = new PermissionListener() {
+            @Override
+            public boolean onRequestPermissionsResult(int i, String[] strings, int[] results) {
+
+                if (results.length > 0) {
+
+                    if (results[0] == PackageManager.PERMISSION_GRANTED
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:109]: Can you please do the check in a loop so we don't hardcode it to 2 permissions,

File: android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java [hunk 11]
```diff
@@ -95,26 +98,45 @@ private static void doLaunch(Context context, HashMap<String, Object> extraData)
...
-
-                    return true;
                 }
-            };
 
+                return true;
+            }
+        };
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```
Comment (saghul) [inline android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:122]: Isn't the audio permission only required for 34 but not before, and notifications were neded since 30 or something? That suggests we have 3 cases:

- no need for permissions at all
- need for media proyection
- need for media proyection AND microphone

Comment (Calinteodor) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:122]: Notification were needed since 33. Don't you think it would be much simpler to include both for >= 33?

Comment (saghul) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:122]: We can do that yeah.

Comment (saghul) [reply android/sdk/src/main/java/org/jitsi/meet/sdk/JitsiMeetOngoingConferenceService.java:122]: Yeah.

**Review by saghul — APPROVED** [review]
